core/domain/platform_parameter_domain.py

```diff
@@ -35,3 +67,976 @@ class PlatformParameterChange(change_domain.BaseChange):
+            for _, client_type in self._conditions:
+                if client_type not in ALLOWED_CLIENT_TYPES:
+                    raise utils.ValidationError(
+                        'Invalid client type %s, must be one of %s.' % (
+                            client_type, ALLOWED_CLIENT_TYPES))
+        elif self._type == 'app_version':
+            for _, version in self._conditions:
+                if not APP_VERSION_WITHOUT_HASH_REGEXP.match(version):
+                    raise utils.ValidationError(
+                        'Invalid version expression %s, expected to match'
+                        'regexp %s.' % (
+                            version, APP_VERSION_WITHOUT_HASH_REGEXP))
+
+    def to_dict(self):
+        """Returns a dict representation of the PlatformParameterFilter domain
+        object.
+
+        Returns:
+            dict. A dict mapping of all fields of PlatformParameterFilter
+            object.
+        """
+        return {
+            'type': self._type,
+            'conditions': self._conditions,
+        }
+
+    @classmethod
+    def from_dict(cls, filter_dict):
+        """Returns an PlatformParameterFilter object from a dict.
+
+        Args:
+            filter_dict: dict. A dict mapping of all fields of
+                PlatformParameterFilter object.
+
+        Returns:
+            PlatformParameterFilter. The corresponding PlatformParameterFilter
+            domain object.
+        """
+        return cls(filter_dict['type'], filter_dict['conditions'])
+
+    @classmethod
+    def _match_version_expression(cls, op, value, client_version):
+        """Tries to match the version expression against the client version.
+
+        Args:
+            op: str. The operator for comparison, e.g. '=', '>'.
+            value: str. The version for comparison, e.g. '1.0.1'.
+            client_version: str. The client version, e.g. '1.0.1-3aebf3h'.
+
+        Returns:
+            bool. True if the expression matches the version.
+        """
+        match = APP_VERSION_WITH_HASH_REGEXP.match(client_version)
+        if match:
+            client_version_without_hash = match.group(1)
+
+            is_equal = value == client_version_without_hash
+            is_client_version_smaller = cls._is_first_version_smaller(
+                client_version_without_hash, value)
+            is_client_version_larger = cls._is_first_version_smaller(
+                value, client_version_without_hash
+            )
+            if op == '=':
+                return is_equal
+            elif op == '<':
+                return is_client_version_smaller
+            elif op == '<=':
+                return is_equal or is_client_version_smaller
+            elif op == '>':
+                return is_client_version_larger
+            elif op == '>=':
+                return is_equal or is_client_version_larger
+        return False
+
+    @staticmethod
+    def _is_first_version_smaller(version_a, version_b):
```

Comment:
Why is this static method?

Reply:
(Answered in above thread.)
Changed `_match_version_expression` `_is_first_version_smaller` to instance method for clarity.